tikv/raftstore/worker.go

```diff
@@ -180,14 +182,80 @@ func newWorker(name string, wg *sync.WaitGroup) *worker {
 	}
 }
 
+type splitCheckKeyEntry struct {
+	key       []byte
+	pos       uint64
+	valueSize uint64
+	cf        string
+}
+
 type splitCheckRunner struct {
 	engine          *badger.DB
 	router          *router
 	coprocessorHost *CoprocessorHost
 }
 
 func (r *splitCheckRunner) run(t task) {
-	// TODO: stub
+	spCheckTask := t.data.(*splitCheckTask)
+	region := spCheckTask.region
+	regionId := region.Id
+	startKey := EncStartKey(region)
+	endKey := EncEndKey(region)
+	log.Debugf("executing task: [regionId: %d, startKey: %s, endKey: %s]", regionId,
+		hex.EncodeToString(startKey), hex.EncodeToString(endKey))
+	/// todo, a metrics method ignored here.
+	host := r.coprocessorHost.newSplitCheckerHost(region, r.engine, spCheckTask.autoSplit,
+		spCheckTask.policy)
+	if host.skip() {
+		log.Debugf("skip split check, [regionId : %d]", regionId)
+	}
+	var keys [][]byte
+	var err error
+	switch host.policy() {
+	case pdpb.CheckPolicy_APPROXIMATE:
+		keys, err = host.approximateSplitKeys(region, r.engine)
+		if err == nil {
+			for i, k := range keys {
+				keys[i] = OriginKey(k)
+			}
+			break
+		}
+		log.Errorf("failed to get approximate split key, try scan way: [regionId: %d, err : %v]",
+			regionId, err)
+		fallthrough
+	case pdpb.CheckPolicy_SCAN:
+		keys, err = r.scanSplitKeys(host, region, startKey, endKey)
+		if err != nil {
+			log.Errorf("failed to scan split key: [regionId: %d, err: %v]", regionId, err)
+			return
+		}
+	}
+	if keys != nil && len(keys) != 0 {
```

Comment:
It's safe to call `len(keys)` even if keys is nil.

Reply:
ok, thanks